concordium-consensus/src/Concordium/ImportExport.hs

```diff
@@ -150,90 +254,185 @@ exportDatabaseV3 ::
+    -- as finalization records, if such a file exists.
+    (genIndex, startHeight, blockIndex, lastChunkFilenameM) <-
+        if indexFileExists
             then do
                 indexContents <- readFile indexFile
-                let lastChunkMetadata =
-                        if not $ null indexContents
-                            then splitOn "," . last . lines $ indexContents
-                            else -- if the index file is empty, reexporting from section 0 and block height 1 will be
-                            -- equivalent to exporting from scratch
-                                ["", "0", "1", "1"]
-                let lastChunkGenesisIndex = GenesisIndex . read $ lastChunkMetadata !! 1
-                let lastChunkFirstBlock = BlockHeight . read $ lastChunkMetadata !! 2
-                -- write back the index file without the last line (corresponding to the chunk deleted above)
-                writeFile (indexFile ++ "~") . unlines . init . lines $ indexContents
-                renameFile (indexFile ++ "~") indexFile
-                return (lastChunkGenesisIndex, lastChunkFirstBlock)
+                blockIndex <- case AP.parseOnly parseBlockIndexFile (T.pack indexContents) of
+                    Left err -> fail $ "An error occurred while parsing '" <> indexFile <> "': " <> err
+                    -- normalize the block index, s.t. superfluous genesis hash lines are removed.
+                    Right bi -> do
+                        return $ normalizeBlockIndex bi
+                case blockIndex of
+                    -- we export blocks starting from height 1 because genesis blocks need not be exported.
+                    Empty -> return (0, 1, Empty, Nothing)
+                    -- the following is safe due to normalization of the blockindex.
+                    _ :|> (gd, chunks) -> case sLast chunks of
+                        BlockIndexChunkInfo{..} -> do
+                            -- the filename of the most recently exported chunk
+                            let lastChunkFile = outDir </> T.unpack filename
+                            -- if we overwrite files, back up the most recently exported
+                            -- chunk file so we can restore it if something goes wrong.
+                            when overwriteExistingChunks $ renameFile lastChunkFile $ lastChunkFile <> "~"
+                            -- evict the last chunk from the block index since we want to
+                            -- export blocks starting from the genesis index and the height
+                            -- of the first block contained within it
+                            return
+                                ( genesisIndex,
+                                  blockHeightFirst,
+                                  sInit blockIndex :|> (gd, sInit chunks),
+                                  Just lastChunkFile
+                                )
```

Comment:
```suggestion
                    _ :|> (gd, chunks) -> case chunks of
                        initChunks :|> BlockIndexChunkInfo{..} -> do
                            -- the filename of the most recently exported chunk
                            let lastChunkFile = outDir </> T.unpack filename
                            -- if we overwrite files, back up the most recently exported
                            -- chunk file so we can restore it if something goes wrong.
                            when overwriteExistingChunks $ renameFile lastChunkFile $ lastChunkFile <> "~"
                            -- evict the last chunk from the block index since we want to
                            -- export blocks starting from the genesis index and the height
                            -- of the first block contained within it
                            return
                                ( genesisIndex,
                                  blockHeightFirst,
                                  sInit blockIndex :|> (gd, initChunks),
                                  Just lastChunkFile
                                )
```
We should also have an `Empty` case. If normalization guarantees no empty chunks, then this can just be an error. However, I'm not sure that's currently the case.